staging/src/k8s.io/apiserver/pkg/endpoints/handlers/negotiation/negotiate.go

```diff
@@ -263,6 +277,26 @@ func acceptMediaTypeOptions(params map[string]string, accepts *AcceptedMediaType
 	return options, true
 }
 
+type candidateMediaType struct {
+	accepted *AcceptedMediaType
+	clauses  goautoneg.Accept
+	length   int
+}
+
+type candidateMediaTypeSlice []candidateMediaType
+
+func (s candidateMediaTypeSlice) Len() int           { return len(s) }
+func (s candidateMediaTypeSlice) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+func (s candidateMediaTypeSlice) Less(i, j int) bool { return s[i].length < s[j].length }
+
+func mediaTypeLength(val goautoneg.Accept) int {
```

Comment:
Not sure if the sheer string length is the correct measure of "being specific".
* What if we ever have a `Type` or `SubType` that's a single character so that they are of the same length as "*"?
* If a clause has more parameters but shorter total length than the other, is it considered to be more specific or less specific?

Reply:
The only thing that should bias priority is the `q` field (handled by the existing code) and the order specified by the user.  We should never sort anything else in order.  Candidates should be evaluated in the order, not sorted.

Reply:
Hi @caesarxuchao @smarterclayton thanks for the comments, left a question here:

I just revisited the code and rfc7231, `q` field already handled well by `goautoneg.ParseAccept` but for scenario: In the absence of `q` field (or all `q` fields are equal) it should follow https://tools.ietf.org/html/rfc7231#section-5.3.2
```
   Media ranges can be overridden by more specific media ranges or
   specific media types.  If more than one media range applies to a
   given type, the most specific reference has precedence.  For example,

     Accept: text/*, text/plain, text/plain;format=flowed, */*

   have the following precedence:

   1.  text/plain;format=flowed

   2.  text/plain

   3.  text/*

   4.  */*
```
So I think if we strictly follow the rfc also not breaking the already sorted `q` field [here](https://github.com/shiywang/kubernetes/blob/3c0c5ed4e019f2c1aef520a7d1f78763368bede7/vendor/bitbucket.org/ww/goautoneg/autoneg.go#L133), we should change the implementation of `ParseAccept` to return most specific value if `q` fields were equal?
Also @caesarxuchao I was wrong about sheer string, maybe we should measure by numbers of Params ?

Reply:
I agree with @shiywang 's interpretation. It seems we should send a patch to goautogen to fix its [accept_slice.Less](https://github.com/shiywang/kubernetes/blob/3c0c5ed4e019f2c1aef520a7d1f78763368bede7/vendor/bitbucket.org/ww/goautoneg/autoneg.go#L64-L77).